File: src/agents/shared/jira.ts
import axios from 'axios';
import fs from 'fs-extra';
import FormData from 'form-data';

type JiraApiErrorData = {
  errorMessages?: string[];
  errors?: Record<string, string>;
};

type CreatedJiraIssue = {
  key: string;
};

type JiraPermission = {
  havePermission: boolean;
};

type JiraIssuePayload = {
  fields: {
    project: {
      id?: string;
      key?: string;
    };
    issuetype: {
      name: string;
    };
    summary: string;
    description: ReturnType<typeof toJiraDescription>;
    labels: string[];
  };
};

type JiraUser = {
  accountId: string;
  displayName?: string;
  emailAddress?: string;
};

type JiraIssue = {
  key: string;
  fields: {
    summary?: string;
    description?: unknown;
    updated: string;
    status?: unknown;
    labels?: string[];
  };
};

function jiraClient() {
  const baseURL = process.env.JIRA_BASE_URL;
  const email = process.env.JIRA_EMAIL;
  const token = process.env.JIRA_API_TOKEN;

  if (!baseURL || !email || !token) {
    throw new Error('Missing Jira environment variables');
  }

  return axios.create({
    baseURL,
    auth: {
      username: email,
      password: token
    },
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  });
}

export function getJiraEnvSummary() {
  return {
    baseURL: process.env.JIRA_BASE_URL ?? '',
    email: process.env.JIRA_EMAIL ?? '',
    hasToken: Boolean(process.env.JIRA_API_TOKEN),
    tokenLength: process.env.JIRA_API_TOKEN?.length ?? 0,
    projectKey: process.env.JIRA_PROJECT_KEY ?? '',
    projectId: process.env.JIRA_PROJECT_ID ?? '',
    issueType: process.env.JIRA_ISSUE_TYPE ?? ''
  };
}

function toJiraDescription(text: string) {
  return {
    type: 'doc',
    version: 1,
    content: text.split('\n').map((line) => ({
      type: 'paragraph',
      content: line
        ? [
            {
              type: 'text',
              text: line.replace(/^\*\*(.*)\*\*$/, '$1'),
              marks: /^\*\*.*\*\*$/.test(line)
                ? [
                    {
                      type: 'strong'
                    }
                  ]
                : undefined
            }
          ]
        : []
    }))
  };
}

function getJiraErrorMessage(error: unknown) {
  if (!axios.isAxiosError(error)) {
    return error instanceof Error ? error.message : String(error);
  }

  const cause = error.cause;
  const responseStatus =
    error.response?.status ??
    (axios.isAxiosError(cause) ? cause.response?.status : undefined);

  if (responseStatus === 401) {
    return (
      'Jira authentication failed with status 401. ' +
      'Check JIRA_EMAIL and JIRA_API_TOKEN, revoke any exposed token, and create a fresh Atlassian API token.'
    );
  }

  const responseData =
    error.response?.data ??
    (axios.isAxiosError(cause) ? cause.response?.data : undefined);
  const data = responseData as JiraApiErrorData | undefined;
  const messages = [
    ...(data?.errorMessages ?? []),
    ...Object.entries(data?.errors ?? {}).map(([field, message]) => `${field}: ${message}`)
  ];

  return messages.length > 0
    ? messages.join('; ')
    : `Jira API request failed with status ${responseStatus ?? 'unknown'}`;
}

async function jiraRequest<T>(request: () => Promise<{ data: T }>): Promise<T> {
  try {
    const response = await request();
    return response.data;
  } catch (error) {
    throw new Error(getJiraErrorMessage(error));
  }
}

export async function verifyJiraAuth(): Promise<JiraUser> {
  const client = jiraClient();
  return jiraRequest<JiraUser>(() => client.get('/rest/api/3/myself'));
}

function sanitizeLabel(label: string) {
  return label.replace(/[^a-zA-Z0-9_-]/g, '-').slice(0, 255);
}

function buildCreateIssuePayload(input: {
  projectId?: string;
  projectKey?: string;
  issueType: string;
  summary: string;
  description: string;
  fingerprint: string;
  labels?: string[];
}): JiraIssuePayload {
  return {
    fields: {
      project: input.projectId
        ? { id: input.projectId }
        : { key: input.projectKey },
      issuetype: {
        name: input.issueType
      },
      summary: input.summary,
      description: toJiraDescription(input.description),
      labels: [
        'playwright-auto',
        `fp-${input.fingerprint.slice(0, 12)}`,
        ...(input.labels ?? [])
      ].map(sanitizeLabel)
    }
  };
}

export async function validateJiraAccess(): Promise<void> {
  const client = jiraClient();
  const projectKey = process.env.JIRA_PROJECT_KEY;
  const projectId = process.env.JIRA_PROJECT_ID;

  if (!projectKey && !projectId) {
    throw new Error('Missing JIRA_PROJECT_KEY or JIRA_PROJECT_ID');
  }

  const projectIdOrKey = projectId ?? projectKey;

  await jiraRequest(() => client.get('/rest/api/3/myself'));

  try {
    await jiraRequest(() => client.get(`/rest/api/3/project/${projectIdOrKey}`));
  } catch (error) {
    throw new Error(
      `Jira project ${projectIdOrKey} is not visible to this API user. ` +
        'Check JIRA_PROJECT_KEY/JIRA_PROJECT_ID and project access.'
    );
  }

  const permissionData = await jiraRequest<{ permissions: Record<string, JiraPermission> }>(() =>
    client.get('/rest/api/3/mypermissions', {
      params: {
        projectKey: projectKey && !projectId ? projectKey : undefined,
        projectId,
        permissions: 'BROWSE_PROJECTS,CREATE_ISSUES'
      }
    })
  );

  const canBrowse = permissionData.permissions.BROWSE_PROJECTS?.havePermission;
  const canCreate = permissionData.permissions.CREATE_ISSUES?.havePermission;

  if (!canBrowse || !canCreate) {
    throw new Error(
      `Jira API user lacks permission for project ${projectIdOrKey}. ` +
        `BROWSE_PROJECTS=${Boolean(canBrowse)}, CREATE_ISSUES=${Boolean(canCreate)}.`
    );
  }
}

export async function searchJiraIssueByFingerprint(fingerprint: string) {
  const client = jiraClient();

  const jql = `labels = "fp-${fingerprint.slice(0, 12)}" AND statusCategory != Done`;

  const data = await jiraRequest<{ issues?: Array<{ key: string }> }>(() =>
    client.post('/rest/api/3/search/jql', {
      jql,
      maxResults: 1,
      fields: ['key', 'summary', 'status']
    })
  );

  return data.issues?.[0];
}

export async function createJiraBug(input: {
  summary: string;
  description: string;
  fingerprint: string;
  labels?: string[];
}): Promise<CreatedJiraIssue> {
  const client = jiraClient();

  const projectKey = process.env.JIRA_PROJECT_KEY;
  const projectId = process.env.JIRA_PROJECT_ID;
  const issueTypes = process.env.JIRA_ISSUE_TYPE
    ? [process.env.JIRA_ISSUE_TYPE]
    : ['Task', 'Bug'];

  if (!projectKey && !projectId) {
    throw new Error('Missing JIRA_PROJECT_KEY or JIRA_PROJECT_ID');
  }

  const errors: string[] = [];

  for (const issueType of issueTypes) {
    try {
      return await jiraRequest<CreatedJiraIssue>(() =>
        client.post(
          '/rest/api/3/issue',
          buildCreateIssuePayload({
            projectId,
            projectKey,
            issueType,
            summary: input.summary,
            description: input.description,
            fingerprint: input.fingerprint,
            labels: input.labels
          })
        )
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      errors.push(`${issueType}: ${message}`);
    }
  }

  throw new Error(`Unable to create Jira issue. ${errors.join(' | ')}`);
}

export async function updateJiraIssueSummary(issueKey: string, summary: string) {
  const client = jiraClient();

  await jiraRequest(() =>
    client.put(`/rest/api/3/issue/${issueKey}`, {
      fields: {
        summary
      }
    })
  );
}

export async function addJiraComment(issueKey: string, text: string) {
  const client = jiraClient();

  await jiraRequest(() =>
    client.post(`/rest/api/3/issue/${issueKey}/comment`, {
      body: toJiraDescription(text)
    })
  );
}

export async function getJiraIssue(issueKey: string): Promise<JiraIssue> {
  const client = jiraClient();

  return jiraRequest<JiraIssue>(() =>
    client.get(`/rest/api/3/issue/${issueKey}`, {
      params: {
        fields: 'summary,description,updated,status,labels'
      }
    })
  );
}

export async function addJiraIssueComment(issueKey: string, text: string) {
  return addJiraComment(issueKey, text);
}

export async function attachFileToJira(issueKey: string, filePath?: string) {
  if (!filePath || !(await fs.pathExists(filePath))) {
    return;
  }

  const baseURL = process.env.JIRA_BASE_URL;
  const email = process.env.JIRA_EMAIL;
  const token = process.env.JIRA_API_TOKEN;

  if (!baseURL || !email || !token) {
    throw new Error('Missing Jira environment variables');
  }

  const form = new FormData();
  form.append('file', fs.createReadStream(filePath));

  await jiraRequest(() =>
    axios.post(
      `${baseURL}/rest/api/3/issue/${issueKey}/attachments`,
      form,
      {
        auth: {
          username: email,
          password: token
        },
        headers: {
          ...form.getHeaders(),
          'X-Atlassian-Token': 'no-check'
        }
      }
    )
  );
}

File: python_agents/agents/shared/jira.py
from __future__ import annotations

import os
from typing import Any

from .http_client import BasicAuthJsonClient, HttpError


def jira_env_summary() -> dict[str, Any]:
    return {
        "baseUrl": os.environ.get("JIRA_BASE_URL", ""),
        "email": os.environ.get("JIRA_EMAIL", ""),
        "tokenConfigured": bool(os.environ.get("JIRA_API_TOKEN")),
        "tokenLength": len(os.environ.get("JIRA_API_TOKEN", "")),
        "projectKey": os.environ.get("JIRA_PROJECT_KEY", ""),
        "projectId": os.environ.get("JIRA_PROJECT_ID", ""),
        "issueType": os.environ.get("JIRA_ISSUE_TYPE", ""),
    }


def jira_configured(require_project: bool = False) -> bool:
    required = ["JIRA_BASE_URL", "JIRA_EMAIL", "JIRA_API_TOKEN"]
    has_base = all(os.environ.get(key) for key in required)
    if require_project:
        return has_base and bool(os.environ.get("JIRA_PROJECT_KEY") or os.environ.get("JIRA_PROJECT_ID"))
    return has_base


class JiraClient:
    def __init__(self) -> None:
        if not jira_configured():
            raise RuntimeError("Missing Jira environment variables")

        self.base_url = os.environ["JIRA_BASE_URL"].rstrip("/")
        self.client = BasicAuthJsonClient(
            self.base_url,
            os.environ["JIRA_EMAIL"],
            os.environ["JIRA_API_TOKEN"],
        )

    def verify_auth(self) -> dict[str, Any]:
        return self.client.request("GET", "/rest/api/3/myself")

    def validate_access(self) -> None:
        project_id = os.environ.get("JIRA_PROJECT_ID")
        project_key = os.environ.get("JIRA_PROJECT_KEY")
        project_id_or_key = project_id or project_key
        if not project_id_or_key:
            raise RuntimeError("Missing JIRA_PROJECT_KEY or JIRA_PROJECT_ID")

        self.verify_auth()
        self.client.request("GET", f"/rest/api/3/project/{project_id_or_key}")
        permissions = self.client.request("GET", "/rest/api/3/mypermissions", params={
            "projectKey": project_key if project_key and not project_id else None,
            "projectId": project_id,
            "permissions": "BROWSE_PROJECTS,CREATE_ISSUES",
        })
        permission_map = permissions.get("permissions", {})
        can_browse = permission_map.get("BROWSE_PROJECTS", {}).get("havePermission")
        can_create = permission_map.get("CREATE_ISSUES", {}).get("havePermission")
        if not can_browse or not can_create:
            raise RuntimeError(
                f"Jira user lacks project permissions. BROWSE_PROJECTS={bool(can_browse)}, "
                f"CREATE_ISSUES={bool(can_create)}"
            )

    def get_issue(self, issue_key: str) -> dict[str, Any]:
        return self.client.request(
            "GET",
            f"/rest/api/3/issue/{issue_key}",
            params={"fields": "summary,description,updated,status,labels"},
        )

    def add_comment(self, issue_key: str, text: str) -> None:
        self.client.request("POST", f"/rest/api/3/issue/{issue_key}/comment", {
            "body": to_jira_adf(text),
        })

    def search_by_fingerprint(self, fingerprint: str) -> dict[str, Any] | None:
        data = self.client.request("POST", "/rest/api/3/search/jql", {
            "jql": f'labels = "fp-{fingerprint[:12]}" AND statusCategory != Done',
            "maxResults": 1,
            "fields": ["key", "summary", "status"],
        })
        issues = data.get("issues") or []
        return issues[0] if issues else None

    def create_bug(self, summary: str, description: str, fingerprint: str, labels: list[str]) -> dict[str, Any]:
        project_key = os.environ.get("JIRA_PROJECT_KEY")
        project_id = os.environ.get("JIRA_PROJECT_ID")
        if not project_key and not project_id:
            raise RuntimeError("Missing JIRA_PROJECT_KEY or JIRA_PROJECT_ID")

        issue_types = [os.environ["JIRA_ISSUE_TYPE"]] if os.environ.get("JIRA_ISSUE_TYPE") else ["Task", "Bug"]
        errors: list[str] = []

        for issue_type in issue_types:
            payload = {
                "fields": {
                    "project": {"id": project_id} if project_id else {"key": project_key},
                    "issuetype": {"name": issue_type},
                    "summary": summary,
                    "description": to_jira_adf(description),
                    "labels": [_sanitize_label(label) for label in ["playwright-auto", f"fp-{fingerprint[:12]}", *labels]],
                }
            }
            try:
                return self.client.request("POST", "/rest/api/3/issue", payload)
            except HttpError as error:
                errors.append(f"{issue_type}: {error}")

        raise RuntimeError("Unable to create Jira issue. " + " | ".join(errors))

    def update_summary(self, issue_key: str, summary: str) -> None:
        self.client.request("PUT", f"/rest/api/3/issue/{issue_key}", {"fields": {"summary": summary}})

    def attach_file(self, issue_key: str, file_path: str | None) -> None:
        if not file_path:
            return
        self.client.upload_file(
            f"/rest/api/3/issue/{issue_key}/attachments",
            file_path,
            {"X-Atlassian-Token": "no-check"},
        )


def verify_jira_auth() -> dict[str, Any]:
    return JiraClient().verify_auth()


def validate_jira_access() -> None:
    JiraClient().validate_access()


def search_jira_issue_by_fingerprint(fingerprint: str) -> dict[str, Any] | None:
    return JiraClient().search_by_fingerprint(fingerprint)


def create_jira_bug(summary: str, description: str, fingerprint: str, labels: list[str] | None = None) -> dict[str, Any]:
    return JiraClient().create_bug(summary, description, fingerprint, labels or [])


def update_jira_issue_summary(issue_key: str, summary: str) -> None:
    JiraClient().update_summary(issue_key, summary)


def add_jira_comment(issue_key: str, text: str) -> None:
    JiraClient().add_comment(issue_key, text)


def get_jira_issue(issue_key: str) -> dict[str, Any]:
    return JiraClient().get_issue(issue_key)


def add_jira_issue_comment(issue_key: str, text: str) -> None:
    add_jira_comment(issue_key, text)


def attach_file_to_jira(issue_key: str, file_path: str | None) -> None:
    JiraClient().attach_file(issue_key, file_path)


def to_jira_adf(text: str) -> dict[str, Any]:
    content = []
    for line in text.splitlines():
        is_bold = line.startswith("**") and line.endswith("**") and len(line) > 4
        clean_text = line[2:-2] if is_bold else line
        content.append({
            "type": "paragraph",
            "content": [
                {
                    "type": "text",
                    "text": clean_text,
                    **({"marks": [{"type": "strong"}]} if is_bold else {}),
                }
            ] if clean_text else [],
        })
    return {"type": "doc", "version": 1, "content": content}


def _sanitize_label(label: str) -> str:
    return "".join(char if char.isalnum() or char in "_-" else "-" for char in label)[:255]
